### ISSUE TITLE
DavMail stops working, `/usr/bin/davmail` prints garbage to the terminal

### ISSUE THREAD

**Issue by ernstki** [mguessan/davmail#376]
To be clear, this was a self-inflicted problem, and I have a resolution. Documenting here for posterity.

## Environment

* systemd-based Linux distro (Ubuntu 20.04-based, elementaryOS 6.1 Jólnir)
* OpenJDK 11 from default distro repository
* DavMail 5.x from default distro repository

…or any later version.

This is a Linux-specific and probably _Debian/Ubuntu-specific_ problem, and **had nothing at all to do with DavMail itself**, or which version of Java was used.

## Symptoms

* DavMail does not launch from the system menus anymore, _i.e._, using the `.desktop` file provided by the distro package
* running `/usr/bin/davmail` (itself a symlink to `/usr/share/davmail/davmail.jar` in Debian/Ubunutu) prints garbage characters to the terminal:
    ```console
    $ davmail |& head
    /usr/bin/davmail: line 1: $'PK\003\004': command not found
    /usr/bin/davmail: line 2: $'\bPTtO': command not found
    /usr/bin/davmail: line 2: $'\312\023': command not found
    /usr/bin/davmail: line 3:PTtOcom/PK: No such file or directory
    /usr/bin/davmail: line 4:PTtcom/ctc/PK: No such file or directory
    com/ctc/wstx/PK: No such file or directory
    /usr/bin/davmail: line 6:PTtOcom/ctc/wstx/sr/PK: No such file or directory
    /usr/bin/davmail: line 7: $'T\325\3658~\356\333g2I': command not found
    /usr/bin/davmail: line 7PTtO0�:�b=��#com/ctc/wstx/sr/StreamScanner.class��: No such file or directory
    /usr/bin/davmail: line 7: C: command not found
    ````

## Solution

Restart the systemd unit that manages the kernel `binfmt_misc` facility.

```bash
sudo systemctl restart binfmt-support
```

…or reboot.

_Why or how would this service ever **not** be running? I had, in haste, stopped a bunch of systemd services trying to expunge snapd (and all its various tentacles), and was apparently a bit too liberal with `xargs`, is all I can figure._

## Discussion

The Debian package links `/usr/share/davmail/davmail.jar` to `/usr/bin/davmail`. This seemed like an odd choice, since _usually_ you need to run `java -jar /path/to/class.jar`—but that was before I knew about [the `binfmt_misc` kernel facility](https://www.kernel.org/doc/html/latest/admin-guide/binfmt-misc.html).

The errors printed to the terminal are as a result of 1) `/usr/share/davmail/davmail.jar` being marked executable; 2) the shell, absent a [shebang line](https://en.wikipedia.org/wiki/Shebang_(Unix)) assuming that its contents are shell script; and 3) attempting to run a binary ZIP file, line-by-line, as a shell script.

Running `hexdump` helps crystallize this notion:

```console
$ head -c 128 /usr/share/davmail/davmail.jar | hexdump -C
00000000  50 4b 03 04 0a 00 00 08  00 00 50 54 74 4f 00 00  |PK........PTtO..|
00000010  00 00 00 00 00 00 00 00  00 00 09 00 04 00 4d 45  |..............ME|
00000020  54 41 2d 49 4e 46 2f fe  ca 00 00 50 4b 03 04 14  |TA-INF/....PK...|
00000030  00 00 00 08 00 50 54 74  4f 7c 39 bd df 95 01 00  |.....PTtO|9.....|
00000040  00 fc 03 00 00 14 00 00  00 4d 45 54 41 2d 49 4e  |.........META-IN|
00000050  46 2f 4d 41 4e 49 46 45  53 54 2e 4d 46 85 93 41  |F/MANIFEST.MF..A|
00000060  6f db 30 0c 85 ef fa 15  3a 6e 18 24 c7 59 8a 01  |o.0.....:n.$.Y..|
00000070  b9 75 1d 30 ec 90 61 87  ad 77 da a6 6d 25 b2 64  |.u.0..a..w..m%.d|
00000080
```

Looks familiar, right?

### What _is_ the binfmt_misc kernel facility?

Quoting from [its documentation](https://www.kernel.org/doc/html/latest/admin-guide/binfmt-misc.html):

> #### Kernel Support for miscellaneous Binary Formats (binfmt_misc)
> 
> This Kernel feature allows you to invoke almost (for restrictions see below) every program by simply typing its name in the shell. This includes for example compiled Java(TM), Python or Emacs programs.
> 
> To achieve this you must tell binfmt_misc which interpreter has to be invoked with which binary. Binfmt_misc recognises the binary-type by matching some bytes at the beginning of the file with a magic byte sequence (masking out specified bits) you have supplied. Binfmt_misc can also recognise a filename extension aka .com or .exe.

## References

* https://stackoverflow.com/questions/1667830/running-a-jar-file-without-directly-calling-java/1667855 - _which gave me the clue that there was a kernel service making it so the `.jar` could run as an ordinary executable, sans the usual `java -jar` that would be otherwise required_
* https://www.kernel.org/doc/html/v6.7/admin-guide/java.html - _a discussion of using the binfmt_misc kernel facility for Java JARs_

**Comment by ernstki** [comment]
Closing as stated above, as this is not a DavMail bug.

**Comment by esabol** [comment]
@ernstki wrote:
> DavMail 5.x from default distro repository

I also just noticed that that's a pretty old version! The current released version is 6.2.2.

**Comment by ernstki** [comment]
> [@ernstki](https://github.com/ernstki) : `/usr/bin/davmail` should be a bash script. It sets a bunch of Java options that need to be set. You can download the latest version at https://raw.githubusercontent.com/mguessan/davmail/refs/heads/master/src/bin/davmail

@esabol Yes, but on recent-ish Ubuntu and Debian, it's not. It's a symlink to a `.jar`, as explained above. The Debian packagers made that decision, and it works fine, except when the user (me) does something to mess it up.

That's why I filed and then immediately closed this issue. To help someone else down the road, if possible, but not create another headache for Mickaël.
